alf/environments/parallel_environment.cpp

```diff
@@ -0,0 +1,501 @@
+      Job job{JobType::reset};
+      job_queues_[original_env_ids_[env_id]]->send(&job, sizeof(job), 0);
+      reseted_[env_id] = true;
+      reset_queue_.push_back(original_env_ids_[env_id]);
+    }
+  }
+  return timestep_array_;
+}
+
+py::object ParallelEnvironment::Reset() {
+  reset_queue_.clear();
+  for (int env_id = 0; env_id < num_envs_ + num_spare_envs_; ++env_id) {
+    Job job{JobType::reset};
+    job_queues_[env_id]->send(&job, sizeof(job), 0);
+    if (env_id < num_envs_) {
+      original_env_ids_[env_id] = env_id;
+      reseted_[env_id] = false;
+    } else {
+      reset_queue_.push_back(env_id);
+    }
+  }
+  reseted_.assign(num_envs_, false);
+  return Step(action_buffer_.ViewAsNestedArray(0, num_envs_));
+}
+
+class ProcessEnvironment : public EnvironmentBase {
+  std::string name_;
+  int env_id_;
+  py::object env_;
+  py::function call_handler_;
+  py::object py_step_;
+  py::object py_reset_;
+  bip::message_queue job_queue_;
+  bip::message_queue ready_queue_;
+  py::object reset_result_;
+  bool just_reseted_;
+  int32_t* env_id_buf_;
+
+ public:
+  ProcessEnvironment(py::object env,
+                     py::function call_handler,
+                     int env_id,
+                     int num_envs,
+                     const py::object& action_spec,
+                     const py::object& timestep_spec,
+                     const std::string& name);
+  ~ProcessEnvironment();
+  void Worker();
+
+ protected:
+  void Step(int env_id);
+  void Reset();
+  void SendTimestep(const py::object& timestep, int env_id);
+};
+
+ProcessEnvironment::ProcessEnvironment(py::object env,
+                                       py::function call_handler,
+                                       int env_id,
+                                       int num_envs,
+                                       const py::object& action_spec,
+                                       const py::object& timestep_spec,
+                                       const std::string& name)
+    : EnvironmentBase(num_envs, action_spec, timestep_spec, name),
+      name_(name),
+      env_id_(env_id),
+      env_(env),
+      call_handler_(call_handler),
+      py_step_(env.attr("step")),
+      py_reset_(env.attr("reset")),
+      job_queue_(bip::open_or_create,
+                 (name + ".job_queue." + std::to_string(env_id)).c_str(),
+                 3,
+                 sizeof(Job)),
+      ready_queue_(bip::open_or_create,
+                   (name + ".ready_queue").c_str(),
+                   num_envs,
+                   sizeof(int)),
+      just_reseted_(false) {
+  env_id_buf_ =
+      reinterpret_cast<int32_t*>(timestep_buffer_.GetBufByFieldName("env_id"));
+}
+
+ProcessEnvironment::~ProcessEnvironment() {
+  bip::message_queue::remove(
+      (name_ + ".job_queue." + std::to_string(env_id_)).c_str());
+  bip::message_queue::remove((name_ + ".ready_queue").c_str());
+}
+
+class ProcessEnvironmentCaller {
```

Comment:
I think the name `Caller` is a bit confusing - how about `ProcessEnvironmentProxy` or `ProcessEnvironmentClient` for better readability for the future reader?